google/cloud/security/scanner/audit/forwarding_rule_rules_engine.py

```diff
@@ -0,0 +1,249 @@
+
+        Args:
+            forwarding_rule (ForwardingRule): The one specific forwarding rule
+                to be matched against all white list rules
+            force_rebuild (bool): If True, rebuilds the rule book. This will
+                reload the rules definition file and add the rules to the book.
+
+        Returns:
+            RuleViolation: A rule violation tuple with all data about the
+                forwarding rule that didnt pass a white list
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+        resource_rules = self.rule_book.get_resource_rules()
+        matched = False
+        for rule in resource_rules:
+            matched = rule.find_match(forwarding_rule) or matched
+            if matched:
+                return None
+
+        if not matched:
+            return self.RuleViolation(
+                violation_type='FORWARDING_RULE_VIOLATION',
+                load_balancing_scheme=forwarding_rule. \
+                    load_balancing_scheme,
+                target=forwarding_rule.target,
+                port_range=forwarding_rule.port_range,
+                port=forwarding_rule.ports,
+                ip_protocol=forwarding_rule.ip_protocol,
+                ip_address=forwarding_rule.ip_address,
+                resource_id=forwarding_rule.resource_id,
+                rule_index=len(resource_rules),
+                resource_type=ResourceType.FORWARDING_RULE)
+
+    def add_rules(self, rules):
+        """Add rules to the rule book.
+
+        Args:
+            rules (dict): rule from file to be added to book
+        """
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+class ForwardingRuleRulesBook(bre.BaseRuleBook):
+    """The RuleBook for forwarding rules resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitions
+        """
+        super(ForwardingRuleRulesBook, self).__init__()
+        self.resource_rules_map = {}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book
+
+        Args:
+            rule_defs (dict): list of rules and their index number
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+
+        Raises:
+            InvalidRulesSchemaError: if rule has format error
+        """
+
```

Comment:
nit: please remove extra newline for consistency (looks like you don't have newlines after the docstrings usually?)